{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "411761a8-8c0b-4ea6-afd3-8838394d6a1a",
   "metadata": {},
   "source": [
    "### Modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef82f6ec-6691-42c9-ae9f-793fea89fdcc",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os, pybedtools\n",
    "import numpy as np, pandas as pd\n",
    "from Bio import SeqIO\n",
    "from collections import OrderedDict\n",
    "from Bio.Seq import IUPACData\n",
    "from itertools import product"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "beb579d3-a7dd-4817-a85e-d8053a6f11d7",
   "metadata": {},
   "source": [
    "### Transcript coordination to Genome coordination"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86e00f67-7011-4d6f-838a-8935840fdb27",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "tx_file = '/PATH/TO/TX/TABLE/enst_hg38Tables.tsv' # available in m6ATM repository: data/enst_hg38Tables.tar.gz.\n",
    "ref_tx = '/PATH/TO/TX/REF/GRCh38_rna_ensembl.fa' # reference transcript from ensemble\n",
    "ref_gn = '/PATH/TO/GENOME/REF/hg38.fa' # reference genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a624e638-45b1-44ba-b4c6-6a8c16ffba24",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def tx_to_gn(results, tx_df, ref_dict_gn):\n",
    "    \n",
    "    ### settings\n",
    "    n_kmer = 5\n",
    "    margin = int((n_kmer-1)*0.5) # 5-mer\n",
    "    shift_range = range(-2, 3) # range to be fixed\n",
    "\n",
    "    ### merge with tx_df\n",
    "    results = results.merge(tx_df, how = 'left', left_on = 'transcript', right_on = 'name')\n",
    "    results = results[results.chrom.isin(list(ref_dict_gn.keys()))]\n",
    "\n",
    "    ### gn conversion\n",
    "    results_gn = results.copy()\n",
    "    results_gn['gn'] = results_gn.apply(get_gn_pos, ref_dict_gn = ref_dict_gn, margin = margin, axis = 1)\n",
    "    results_gn['gn_pos'] = [int(i.split('_')[0]) for i in results_gn['gn']]\n",
    "    results_gn['gn_motif'] = [i.split('_')[1] for i in results_gn['gn']]\n",
    "\n",
    "    ### try to fix false_table with shifts\n",
    "    results_true = results_gn.loc[lambda row: row['motif'] == row['gn_motif']]\n",
    "    results_false = results_gn.loc[lambda row: row['motif'] != row['gn_motif']]\n",
    "\n",
    "    results_fixed = try_shift(results_false, ref_dict_gn, shift_range = shift_range, margin = margin)\n",
    "    if len(results_fixed)>0:\n",
    "        results_gn = pd.concat([results_true, results_fixed], axis = 0)\n",
    "    else:\n",
    "        results_gn = results_true\n",
    "\n",
    "    ### add info\n",
    "    results_gn = results_gn.reset_index(drop = True)\n",
    "    results_gn['gn_pos_1'] = [i+1 for i in results_gn['gn_pos']]\n",
    "    results_gn['gn_site'] = [x+'_'+str(y) for x, y in zip(results_gn['chrom'], results_gn['gn_pos_1'])]\n",
    "\n",
    "    results_gn = results_gn.groupby('gn_site', as_index = False).agg({'transcript': lambda x: ','.join(x),\n",
    "                                                                      'position': lambda x: merge_int(x),\n",
    "                                                                      'motif': 'first',\n",
    "                                                                      'probability': 'mean',\n",
    "                                                                      'ratio': 'mean',\n",
    "                                                                      'm6a': lambda x: ','.join(x),\n",
    "                                                                      'name2': 'first',\n",
    "                                                                      'gn_pos': 'first',\n",
    "                                                                      'gn_pos_1': 'first',\n",
    "                                                                      'chrom': 'first', \n",
    "                                                                      'strand': 'first',\n",
    "                                                                      'coverage': 'sum',\n",
    "                                                                      'gn_motif' : 'count'})\n",
    "\n",
    "    results_gn.rename(columns = {'gn_motif': 'tx_count'}, inplace = True)\n",
    "    \n",
    "    return results_gn\n",
    "\n",
    "def get_gn_pos(row, ref_dict_gn, margin):\n",
    "    \n",
    "    chrom = row['chrom']\n",
    "    strand = row['strand']\n",
    "    pos = row['position'] # based on transcriptome: 0-start\n",
    "    \n",
    "    ### exon info \n",
    "    exons_start = [int(i) for i in row['exonStarts'].split(',')[:-1]]\n",
    "    exons_end = [int(i) for i in row['exonEnds'].split(',')[:-1]]\n",
    "\n",
    "    ### tx to gn conversion\n",
    "    exon_len = 0\n",
    "    if strand == '+':\n",
    "        for x, y in zip(exons_start, exons_end):\n",
    "            exon_len_last = exon_len\n",
    "            exon_len += (y-x)\n",
    "            if (pos+1)<=exon_len:\n",
    "                gn_index = x+(pos-exon_len_last)\n",
    "\n",
    "                kmer = ref_dict_gn[chrom][gn_index-margin:gn_index+margin+1].upper()\n",
    "                kmer = ''.join(kmer)\n",
    "                break\n",
    "\n",
    "    if strand == '-':\n",
    "        for x, y in zip(exons_start[::-1], exons_end[::-1]):\n",
    "            exon_len_last = exon_len\n",
    "            exon_len += (y-x)\n",
    "            if (pos+1)<=exon_len:\n",
    "                gn_index = y-(pos-exon_len_last)-1\n",
    "\n",
    "                kmer = ref_dict_gn[chrom][gn_index-margin:gn_index+margin+1].reverse_complement().upper()\n",
    "                kmer = ''.join(kmer)\n",
    "                break\n",
    "    \n",
    "    result = str(gn_index)+'_'+kmer\n",
    "    \n",
    "    return result\n",
    "\n",
    "def try_shift(false_table, ref_dict_gn, shift_range = range(-2,3), margin = 2):\n",
    "    \n",
    "    t_table = []\n",
    "    for shift in shift_range:\n",
    "\n",
    "        table = false_table.copy()\n",
    "        table.loc[:,'gn_pos'] = [int(i+shift) for i in table.gn_pos.tolist()]\n",
    "        table.loc[:,'gn_motif'] = [str(ref_dict_gn[chrom][int(pos)-margin:int(pos)+margin+1].upper()) for chrom, pos in zip(table.chrom, table.gn_pos)]\n",
    "\n",
    "        # true or false\n",
    "        t_idx = [x == y for x, y in zip(table['motif'], table['gn_motif'])]\n",
    "        t_row = table.loc[t_idx,:]\n",
    "\n",
    "        if t_row.shape[0]>0:\n",
    "            t_table.append(t_row)\n",
    "\n",
    "    if len(t_table)>0:\n",
    "        t_table = pd.concat(t_table, axis = 0)\n",
    "\n",
    "    return t_table\n",
    "\n",
    "def merge_int(x):\n",
    "    \n",
    "    str_list = list(map(str, x))\n",
    "    merged = ','.join(str_list)\n",
    "    \n",
    "    return merged\n",
    "\n",
    "def get_ref_dict(ref_path):\n",
    "    \n",
    "    ref_dict = OrderedDict()\n",
    "    for record in SeqIO.parse(ref_path, 'fasta'):\n",
    "        ref_dict[record.id] = record.seq\n",
    "        \n",
    "    return ref_dict\n",
    "\n",
    "def add_kmer(row):\n",
    "    \n",
    "    tx = row['transcript']\n",
    "    pos = row['position']-1\n",
    "    motif = ref_dict_tx[tx][pos-2:pos+3]\n",
    "    motif = ''.join(motif)\n",
    "    \n",
    "    return motif\n",
    "\n",
    "def extend_ambiguous(seq):\n",
    "    \n",
    "    d = IUPACData.ambiguous_dna_values\n",
    "    return list(map(''.join, product(*map(d.get, seq)))) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f0229ec-51f8-4ffd-9b0c-48ff11ade325",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "### load reference files\n",
    "tx_df = pd.read_csv(tx_file, sep = '\\t')\n",
    "tx_df['name'] = [i.split('.')[0] for i in tx_df['name']]\n",
    "ref_dict_tx = get_ref_dict(ref_tx)\n",
    "ref_dict_gn = get_ref_dict(ref_gn)\n",
    "DRACH = extend_ambiguous('DRACH')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10724ec3-1164-4b29-924d-d5946596be28",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "### m6ATM\n",
    "results = pd.read_csv('../data/hek293tx_m6atm.csv')\n",
    "results_gn = tx_to_gn(results, tx_df, ref_dict_gn)\n",
    "\n",
    "### m6Anet\n",
    "pred_m6anet = pd.read_csv('../data/hek293tx_m6anet.csv')\n",
    "pred_m6anet.columns = ['transcript', 'position', 'coverage', 'probability', 'motif', 'ratio']\n",
    "pred_m6anet['m6a'] = '-'\n",
    "pred_m6anet['transcript'] = [i.split('.')[0] for i in pred_m6anet['transcript']]\n",
    "\n",
    "pred_m6anet_gn = tx_to_gn(pred_m6anet, tx_df, ref_dict_gn)\n",
    "\n",
    "### m6aBasecaller\n",
    "pred_m6abase = pd.read_csv('../data/hek293tx_m6abase.tsv', sep = '\\t')\n",
    "pred_m6abase.columns = ['transcript', 'position', 'ref_base', 'strand', 'm6a', 'coverage', 'base_accuracy', 'ratio', 'probability']\n",
    "pred_m6abase = pred_m6abase.loc[:,['transcript', 'position', 'coverage', 'probability', 'ratio', 'm6a', 'ref_base']]\n",
    "pred_m6abase = pred_m6abase[pred_m6abase['ref_base'] == 'A']\n",
    "pred_m6abase['probability'] = pred_m6abase['probability'].replace(np.nan, 0)\n",
    "pred_m6abase['motif'] = pred_m6abase.apply(add_kmer, axis = 1)\n",
    "pred_m6abase = pred_m6abase[pred_m6abase['motif'].isin(DRACH)]\n",
    "pred_m6abase['transcript'] = [i.split('.')[0] for i in pred_m6abase['transcript']]\n",
    "\n",
    "pred_m6abase_gn = tx_to_gn(pred_m6abase, tx_df, ref_dict_gn)\n",
    "\n",
    "### Tombo\n",
    "pred_tombo = pd.read_csv('../data/hek293tx_tombo.bed', sep = '\\t', header = None)\n",
    "pred_tombo.columns = ['transcript', 'start', 'position', 'id', 'probability']\n",
    "pred_tombo['ratio'] = 0\n",
    "pred_tombo['m6a'] = '-'\n",
    "pred_tombo['motif'] = pred_tombo.apply(add_kmer, axis = 1)\n",
    "\n",
    "pred_tombo_bed = pybedtools.BedTool.from_dataframe(df = pred_tombo)\n",
    "coverage_bed = pybedtools.BedTool('../data/hek293_tombo.coverage.plus.bedgraph')\n",
    "pred_tombo_bed = pred_tombo_bed.intersect(coverage_bed, wb = True, wa = True)\n",
    "pred_tombo_bed = pred_tombo_bed.to_dataframe()\n",
    "\n",
    "pred_tombo['coverage'] = pred_tombo_bed['blockStarts']\n",
    "pred_tombo['transcript'] = [i.split('.')[0] for i in pred_tombo['transcript']]\n",
    "\n",
    "pred_tombo_gn = tx_to_gn(pred_tombo, tx_df, ref_dict_gn)\n",
    "\n",
    "### MINES\n",
    "pred_mines = pd.read_csv('../data/hek293tx_mines.bed', sep = '\\t', header = None)\n",
    "pred_mines.columns = ['transcript', 'start', 'position', 'motif', 'id', '_strand', 'probability', 'coverage']\n",
    "pred_mines['probability'] = pred_mines['probability'].replace('.', 0)\n",
    "pred_mines['probability'] = pred_mines['probability'].astype('float')\n",
    "pred_mines['ratio'] = 0\n",
    "pred_mines['m6a'] = '-'\n",
    "pred_mines['transcript'] = [i.split('.')[0] for i in pred_mines['transcript']]\n",
    "\n",
    "pred_mines_gn = tx_to_gn(pred_mines, tx_df, ref_dict_gn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bece5a8a-c47e-453b-ab03-04510740127b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
